src/librustc_typeck/check/method/suggest.rs

```diff
@@ -468,9 +495,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         } else {
             let limit = if candidates.len() == 5 { 5 } else { 4 };
             for (i, trait_did) in candidates.iter().take(limit).enumerate() {
-                msg.push_str(&format!("\ncandidate #{}: `use {};`",
```

Comment:
These should probably become a real suggestion,  in which case you don't need manual limiting anymore, because the renderer takes care of that

Reply:
The suggestions for inserting use statements do that already

Reply:
Suggestions currently do not preserve their position in the output, they always come at then end. I need to change that for the first requested fix.